### PR TITLE
DOC: Add direct C-extension testing guide

### PR DESCRIPTION
**Pull request by ReemHamraz**
## Description

Hi everyone!!

This PR adds `TESTING_EXTENSIONS.rst` to the developer docs and wires it up in `index_dev.rst`.

Over the course of my GSoC project, I've been writing a lot of tests for our C and Cython extensions. Since establishing this direct test layer is a big prerequisite for safely splitting the low-level layer into a separate package (the APE split), I wanted to put together a proper guide on how to test these compiled modules directly and in pure isolation.

Here is a quick rundown of what the guide covers:

-  How to bypass high-level Python wrappers and hit the C-slots directly using minimal shims and `.view()` casting.
-  Why we need to use strict `type()` checks instead of `isinstance()` at the C-boundary, and guidelines for using `numpy.typing.NDArray`.
-  Tips for exhaustive structural validation, using `zip(..., strict=True)`, and handling tricky edge cases like explosive memory allocations and `np.nan`.
-  Best practices for catching and safely validating C-level crashes across the Python boundary using substring matching.
-  Architectural rules for static analysis, like locking dimensions with `TypeVar` and enforcing positional-only arguments.

I'd really love to get your thoughts on this! Please let me know if anything feels off, if the tone works, or if there are any sections that need more clarity. Looking forward to any suggested changes or improvements, I'm more than happy to iterate on this to make sure it's as helpful as possible for future contributors.

## Related Issues

- Relates to [GSoC 2026 Proposal: Hardening Astropy's Core Stability](https://github.com/OpenAstronomy/gsoc-proposals/blob/main/2026/astropy/astropy-hardening-core-stability-reem-hamraz.md)

cc: @neutrinoceros

### CODE REVIEW

File: docs/development/testing_extensions.rst [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+When testing how the C-engine handles truncated byte streams or malformed data, it should safely raise an error across the Python boundary rather than crashing the interpreter.
+
+Use standard Python terminology for exceptions (e.g., the engine "raises" a ``ValueError``). Because C-level errors often prefix unpredictable memory addresses or line numbers, use substring matching rather than exact string comparisons to validate the error:
+
+.. code-block:: python
+
+    import pytest
+
+    with pytest.raises(ValueError) as exc_info:
```
Comment (mhvk) [inline docs/development/testing_extensions.rst:99]: why not just `raises(ValueError, match="no element found")`, which will match on a substring?

File: docs/development/testing_extensions.rst [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+3. Target the Boundary
+----------------------------
+Differentiate between pure C math engines and the Python-toC wrapper. The tests and type stubs should
+specifically target the wrapper boundary where Python objects are unpacked into C memory pointers.
+
+
+Writing the Tests
+=================
+
+Data Structures and Parametrization
```
Comment (mhvk) [inline docs/development/testing_extensions.rst:45]: This part does not feel specific to C extensions.

File: docs/development/testing_extensions.rst [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+Strict Typing and Assertions
+============================
+
+1. Type Assertions (No ``isinstance``)
+--------------------------------------
+Avoid using ``isinstance()`` for return type validation at the C-boundary, as standard array slicing
+can strip subclasses while masked array slicing preserves them.
+Use strict ``type(x) is y`` assertions to verify the exact return type expected from the C-engine.
+
+2. Strict Type Hinting
```
Comment (mhvk) [inline docs/development/testing_extensions.rst:70]: This and the following item are about typing, not testing. I think they may be useful, but maybe the goal should then be something like "Making C extensions usuable and testable"?

**Review by mhvk — COMMENTED** [review]
Thanks, this looks useful! Some in-line comments, of which the main one is that this guide would seem to not just concern itself with testing, so probably all it needs is to change the title and intro, that it is meant to tell how to provide a clearly typed interface, and how to test the C extension in isolation.

But very nice work!

File: docs/development/testing_extensions.rst [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+========================================
+
+Welcome to the low-level testing guide for Astropy's compiled extensions!
+If you're reading this, you are probably about to write tests for C or Cython modules.
+
+While Astropy's public API is heavily tested, our underlying compiled building blocks (the "hot-paths")
+need their own isolated testing. This guide explains how to operate one layer below the public API,
+right at the boundary between Python and compiled code. Doing this ensures our C-engine safely handles
+memory sizing, index mapping, and string parsing on its own—an essential prerequisite for milestones
+like the APE split and making our core architecture rock solid.
```
Comment (neutrinoceros) [inline docs/development/testing_extensions.rst:12]: I suggest dropping references to the APE because it's nowhere near accepted right now (my bad for not making progress on it), and even in the best case scenario it won't make much sense in hindsight.
```suggestion
memory sizing, index mapping, and string parsing on its own—an essential prerequisite for our core architecture rock solid.
```

File: docs/development/testing_extensions.rst [hunk 5]
```diff
@@ -0,0 +1,113 @@
+========================================
+Testing C and Cython Extensions Directly
```
Comment (neutrinoceros) [inline docs/development/testing_extensions.rst:2]: I'd avoid referring to exact programming languages (RIIR anyone ?)
```suggestion
Testing Low-Level Extension Modules Directly
```

File: docs/development/testing_extensions.rst [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+and boolean masking without the Python layer hiding potential regressions.
+
+Think of this test layer as the bridge between the high-level user API and the low-level C code.
+
+Testing Architecture & Setup Guide
+==================================
+
+1. Bypass the APIs
+------------------
+Construct the minimal raw numpy arrays, structs, or byte streams required to trigger the compiled logic directly.
```
Comment (neutrinoceros) [inline docs/development/testing_extensions.rst:29]:
```suggestion
Construct the minimal raw numpy arrays, structs, or byte streams required to call compiled functions.
```

File: docs/development/testing_extensions.rst [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+
+2. Isolate C-slots
+----------------------------
+When testing C-slots (like ``__getitem__``), use minimal shim classes and ``.view()`` casting to map
+directly onto raw memory buffers rather than using standard Astropy objects.
```
Comment (neutrinoceros) [inline docs/development/testing_extensions.rst:34]: "standard" is a bit overloaded. I tend to read it as "non-third-party" or "implementation-independent", neither of which apply here (I guess astropy is technically first-party in its own context, but definitely third-party everywhere else)
```suggestion
directly onto raw memory buffers rather than using Astropy objects.
```

File: docs/development/testing_extensions.rst (hunk 2 above)
Comment (neutrinoceros) [reply docs/development/testing_extensions.rst:45]: agreed. It's definitely innovative in the context of astropy, and championed by your project, but this would be useful in other contexts as well

File: docs/development/testing_extensions.rst [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+Ensure every parametrized test uses explicit IDs via ``pytest.param(..., id="name_of_case")``.
+
+
+Strict Typing and Assertions
+============================
+
+1. Type Assertions (No ``isinstance``)
+--------------------------------------
+Avoid using ``isinstance()`` for return type validation at the C-boundary, as standard array slicing
+can strip subclasses while masked array slicing preserves them.
```
Comment (neutrinoceros) [inline docs/development/testing_extensions.rst:67]: Note to other reviewers that this is indeed most valuable around C-boundaries. In pure Python these checks could be performed at type-check time and wouldn't need runtime testing.

File: docs/development/testing_extensions.rst [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+Use strict ``type(x) is y`` assertions to verify the exact return type expected from the C-engine.
+
+2. Strict Type Hinting
+----------------------
+Do not use generic ``list`` or ``np.ndarray`` type hints. You must use ``numpy.typing.NDArray`` with
+exact 1D shapes and dtypes (e.g., ``np.ndarray[tuple[int], np.dtype[np.intp]]``).
```
Comment (neutrinoceros) [inline docs/development/testing_extensions.rst:73]: this is something type checkers already enforce. Pyrefly, ty, and mypy (strict mode) all warn/error about unspecified generics. Maybe it's still valuable as part of the guide while we don't have continuous type checking and most core devs are not familiar with the type system though.

File: docs/development/testing_extensions.rst [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+
+3. Semantic Hinting
+-------------------
+When passing dictionaries that the C-engine will strictly read and not mutate, use ``collections.abc.Mapping``
+instead of ``dict``. This signals your exact intent to static type checkers.
+
+Data Validation and Edge Cases
+==============================
+
+* **Exhaustive Validation:** Don't just check memory boundaries (e.g., ``assert type(event) is tuple``). Implement deep, strict equality checks to prove the C-engine's internal calculations (like coordinate tuples or state flags) are completely flawless.
```
Comment (neutrinoceros) [inline docs/development/testing_extensions.rst:83]: not sure about using `event` as a placeholder variable name. I think it's very common in web, but we don't actually encounter it this much in a datascience library.

File: docs/development/testing_extensions.rst [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+* **Strict Iterables:** Whenever using ``zip()`` to compare expected versus actual runtime events, you must include ``strict=True`` to prevent the silent truncation of mismatched lists.
```
Comment (neutrinoceros) [inline docs/development/testing_extensions.rst:84]: this is exactly `zip-without-explicit-strict`(https://docs.astral.sh/ruff/rules/zip-without-explicit-strict/), a linting rule we *should* be enforcing globally in my opinion, but are currently disabling for what assume are merely historical reasons (nobody wants to deal with the churn ?). I wouldn't include it in this guide.

File: docs/development/testing_extensions.rst [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+* **Isolate Edge Cases:** Keep standard tests (where clean set-theory math applies) separated from edge cases like Cartesian expansions (which cause explosive O(N^2) memory allocations) or ``np.nan`` identities (which break standard logical assertions).
```
Comment (neutrinoceros) [inline docs/development/testing_extensions.rst:85]: I don't think I get it. What are "standard logical assertions" ?

File: docs/development/testing_extensions.rst [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+* **Memory Efficiency:** If you need a boolean mask array of all ``False``, don't type out ``[False, False]``. Use ``np.full(N, False)`` for better memory efficiency.
```
Comment (neutrinoceros) [inline docs/development/testing_extensions.rst:86]: Feels like the example is a bit odd; I don't think any reviewer would care about memory efficiency for a 2-elements array. I think I would rather just not include this item at all.

File: docs/development/testing_extensions.rst [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+
+Catching C-Level Exceptions
+===========================
+
+When testing how the C-engine handles truncated byte streams or malformed data, it should safely raise an error across the Python boundary rather than crashing the interpreter.
```
Comment (neutrinoceros) [inline docs/development/testing_extensions.rst:91]:
```suggestion
When testing how the C-engine handles untrusted or malformed data, it should safely raise an error across the Python boundary rather than crashing the interpreter.
```

File: docs/development/testing_extensions.rst [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+
+    assert "no element found" in str(exc_info.value).lower()
+
+Type Stubs (.pyi) Architecture
+==============================
+Static type checkers (like Pyright or mypy) cannot easily natively parse ``.pyx`` files.
+We bridge this gap using ``.pyi`` stub files.
+
+* **Locate the True Boundary:** Distinguish between pure C math engines and the actual Python wrapper. We only type the wrapper.
+* **Argument Enforcement:** If a hand-written C-API parser doesn't use a ``kwlist``, it doesn't inherently support keyword arguments. You must use the ``/`` marker in your ``.pyi`` file to enforce positional-only arguments.
```
Comment (neutrinoceros) [inline docs/development/testing_extensions.rst:111]: I'd link to https://peps.python.org/pep-0570/ here, because it's hardly searchable if you don't already know about it

File: docs/development/testing_extensions.rst [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+* **Dimensionality Locking:** If the C-engine expects multiple arrays that must match in dimensions (like an image and a kernel), use bounded ``TypeVar`` aliases mapped to ``np.ndarray`` to mathematically enforce dimensionality matching at the static level.
```
Comment (neutrinoceros) [inline docs/development/testing_extensions.rst:112]:
```suggestion
* **Dimensional Consistency:** If the C-engine expects multiple arrays that must match in dimensions (like an image and a kernel), use bounded ``TypeVar`` aliases mapped to ``np.ndarray`` to mathematically enforce dimensionality matching at the static level.
```

File: docs/development/testing_extensions.rst [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+* **Linter Compliance:** To pass Ruff's private export rules (PYI001), any internal ``TypeVar`` or ``TypeAlias`` used strictly for the stub file must be prefixed with an underscore (e.g., ``_D = TypeVar("_D")``).
```
Comment (neutrinoceros) [inline docs/development/testing_extensions.rst:113]: Again I don't think it's worth re-hashing linting rules in this guide as they already are bound to pop up in CI and review.

File: docs/index_dev.rst [hunk 18]
```diff
@@ -63,6 +63,7 @@ Details
    development/development_details
    development/codeguide
    development/testguide
+   development/TESTING_EXTENSIONS
```
Comment (neutrinoceros) [inline docs/index_dev.rst]: this filename should probably just match the existing naming convention (all lower case)

**Review by neutrinoceros — COMMENTED** [review]
Great initiative Reem ! It's great to have a reference of what you learned as a resources for future contributors in this area ! I do have a couple questions and suggestions, of course :)

**Comment by ReemHamraz** [comment]
Thanks so much for the reviews @mhvk & @neutrinoceros ! I really appreciate the feedback!!

Let me know what you think of this version! Please suggest changes if anything else needs tweaking.

File: docs/development/testing_extensions.rst (hunk 4 above)
Comment (mhvk) [reply docs/development/testing_extensions.rst:12]: Huh, I always liked that APE!? Why would it not make that much sense? Avoiding compilation still seems quite a strong argument.

Comment (neutrinoceros) [reply docs/development/testing_extensions.rst:12]: I mean that if the APE is accepted and implemented, this sentence would read more and more like an historical remnant as time goes by. There's room for rephrasing it so that the argument is preserved but it reads equally relevant in the future.

**Review by neutrinoceros — APPROVED** [review]
LGTM now. If @mhvk is alright with it I suggest we just get this one in. Thanks Reem !

File: docs/development/testing_extensions.rst (hunk 9 above)
Comment (mhvk) [inline docs/development/testing_extensions.rst:73]: Here, you tell to use `numpy.typing.NDArray`, but then give an example with `np.ndarray` (I just learned that `NDArray` is different, in not needing the shape and `np.dtype` part). Anyway, I'd make the sentence consistent.

File: docs/development/testing_extensions.rst (hunk 13 above)
Comment (mhvk) [reply docs/development/testing_extensions.rst:86]: Unresolving since to me this feels weird too, and it has nothing to do with testing C extensions. I'd definitely remove it.

File: docs/development/testing_extensions.rst (hunk 1 above)
Comment (mhvk) [reply docs/development/testing_extensions.rst:99]: I'm not sure why my comment was marked as resolved when it was not in fact done? I still think one should do `match="no element found"` here.

**Review by mhvk — COMMENTED** [review]
A few more suggestions, and a general worry: it looks like comments have been resolved without actually applying the suggested changes, or commenting on why they have not been applied. Please go through all of them carefully.

Overall, I still feel much of this is general not specific to C extensions, and very focused on typing, without explaining why that is so important (I can sort-of see how it is more relevant than pure python because one cannot as easily just inspect the code to see what is happening, but why not state that?)
